onnxruntime/contrib_ops/rocm/bert/fast_gelu_impl.cu

```diff
@@ -60,153 +59,118 @@ __global__ void FastGeluKernel(const T a, const T b, const T c, const T oneT, co
-__global__ void FastGeluKernel4(const half2 a, const half2 b, const half2 c, const half2 one2, const half2 two2,
-                                int input_length, const float2* input, float2* output) {
-  const int idx = blockIdx.x * TPB + threadIdx.x;
-
-  if (idx < input_length) {
-    float2 input_vec = input[idx];
-    float2 output_vec = output[idx];
-
-    half2* input_half = reinterpret_cast<half2*>(&input_vec);
-    half2* output_half = reinterpret_cast<half2*>(&output_vec);
-
-    half2 lo_data = input_half[0];
-    half2 hi_data = input_half[1];
-
-    const half2 lo_u = two2 * lo_data * (c * lo_data * lo_data + b);
-    const half2 hi_u = two2 * hi_data * (c * hi_data * hi_data + b);
-    const half2 lo_emu = h2exp(-lo_u);
-    const half2 hi_emu = h2exp(-hi_u);
-    const half2 lo_cdf = a + a * (two2/(one2 + lo_emu) - one2);
-    const half2 hi_cdf = a + a * (two2/(one2 + hi_emu) - one2);
-
-    output_half[0] = lo_data * lo_cdf;
-    output_half[1] = hi_data * hi_cdf;
+    using VecT = aligned_vector<T, ILP>;
+    T input_v[ILP];
+    VecT* input_val = reinterpret_cast<VecT*>(&input_v);
+    *input_val = *reinterpret_cast<const VecT*>(&input[idx]);
+    T output_v[ILP];
+    VecT* output_val = reinterpret_cast<VecT*>(&output_v);
+    T bias_v[ILP];
+    if (bias != nullptr) {
+        VecT* bias_val = reinterpret_cast<VecT*>(&bias_v);
+        *bias_val = *reinterpret_cast<const VecT*>(&bias[idx % bias_length]);
+    }
 
-    output[idx] = output_vec;
+    #pragma unroll
+    for (int i = 0; i < ILP; i++) {
+      const T x = (bias == nullptr) ? input_v[i] : input_v[i] + bias_v[i];
+      const T u = twoT * x * (c * x * x + b);
+      const T emu = __expf(-u);
+      const T cdf = a + a * (twoT/(oneT + emu) - oneT);
+      output_v[i] = x * cdf;
+    }
+    *(reinterpret_cast<VecT*>(&output[idx])) = *reinterpret_cast<VecT*>(&output_v[0]);
   }
 }
 
 template <>
 bool LaunchFastGeluKernel(const hipDeviceProp_t& prop, hipStream_t stream, int input_length, int bias_length,
                           const float* input, const float* bias, float* output, bool /*use_half2*/) {
-  constexpr int blockSize = 256;
-  const int gridSize = (input_length + blockSize - 1) / blockSize;
-  hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel<float, blockSize>), dim3(gridSize), dim3(blockSize), 0,
-                     stream, A, B, C, one, two, input_length, bias_length, input, bias, output);
-
+  constexpr int block_size = 256;
+  const int grid_size = (input_length + block_size - 1) / block_size;
+  hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel<float, block_size>), dim3(grid_size), dim3(block_size), 0,
+                     stream, input_length, bias_length, input, bias, output);
   return HIP_CALL(hipPeekAtLastError());
 }
 
 template <>
 bool LaunchFastGeluKernel(const hipDeviceProp_t& prop, hipStream_t stream, int input_length, int bias_length,
                           const half* input, const half* bias, half* output, bool use_half2) {
```

Comment:
use_half2 not needed any more?

Reply:
It is defined in [fast_gelu.cc](https://github.com/microsoft/onnxruntime/blob/07998575a4da229b7b9822254ceda9db6dfdcae7/onnxruntime/contrib_ops/cuda/bert/fast_gelu.cc#L35). I am not sure if it is because users might want to turn this on and off depending on their purposes. 
@tianleiwu do you think we can remove it?

Reply:
Thanks, I think we can keep it then.